### ISSUE TITLE
OpenMM 7.6 release issues and next steps

### ISSUE THREAD

**Issue by jchodera** [openmm/openmm#3232]
Given the number of issues that have appeared following the OpenMM 7.6 release, I wanted to collect the major ones in this thread and discuss short, medium, and long-term paths forward to address the issues that have arisen.

These are major issues that have been reported to make downstream code stop working with the OpenMM 7.6 release, for various reasons:
* ([Parmed #1185](https://github.com/ParmEd/ParmEd/issues/1185)): All released versions of [ParmEd](https://github.com/parmed/parmed) are incompatible with OpenMM 7.6, which breaks all packages that use ParmEd. ParmEd is highly utilized in downstream packages since it is a key interoperability component of the open source molecular simulation ecosystem, with [119K direct conda downloads](https://anaconda.org/conda-forge/parmed/badges) and [307K downloads through `ambertools`](https://anaconda.org/conda-forge/ambertools/badges), so this is a major incompatibility issue.
* https://github.com/openmm/openmm/pull/3046 removed a number of `Amoeba*Force` terms from the C++ and Python APIs without providing backwards-compatible replacements, breaking any code that relied on these forces. This was not [clearly communicated in the release notes](https://github.com/openmm/openmm/releases/tag/7.6.0).
* (https://github.com/openmm/openmm/issues/3218) ([openmmtools #173](https://github.com/openmm/openmmforcefields/pull/173)) : Namespace changes broke some packages that were using parts of the OpenMM internal namespace, or even just `from openmm.app.element import Element` because, even though this wasn't in the internal namespace, our hack to preserve "legacy" imports is incomplete.

Unfortunately, despite an extensive release beta and candidate period, we did not discover most of these issues until after the OpenMM 7.6 release when many things started to break. This is perhaps partly due to the very low download numbers for the beta and release candidates, which are generally not incorporated into continuous integration nightly tests to warn of problems.

This thread is to discuss what we do next to learn from these issues.

# Short term

Since we have already released OpenMM 7.6, if a small number of users are already using this package successfully, it could potentially be problematic for reproducibility to remove it entirely. It has, however, only been out for a few days. The simplest fix may be to request the conda-forge core team to move the `openmm-7.6.0` package to a different label, such as `rc`, `broken`, or `testing` until we can get a 7.6.1 patch out that addresses these issues.

As I type, I see that [this patch to ParmEd to constrain `openmm < 7.6` for `parmed < 3.4.3` may also remedy one of these issues.

# Medium term

Since we have elected not to wage a major campaign to communicate the API-breaking nature of 7.6.0 sufficiently in advance to allow everyone to update their software, it may be sensible to try to restore backwards-compatibility as much as possible to enable code to "just work" unmodified, including the implementation of Python or C++ wrappers for the removed Amoeba forces, along with deprecation warnings that these will be removed fully in OpenMM 8.0.

# Longer term

We need a more robust beta and release candidate feedback mechanism. It would be optimal if we could work with key stakeholders in the OpenMM ecosystem of downstream packages to integrate automated nightly tests of nightly builds and beta/release candidates so we can have an "early warning system" of any changes that will impact major stakeholders, allowing us ample time to work with them prior to a release that will cause major downstream issues.

I know there are strongly-held opinions on semantic versioning, but it is a widely used convention that aligns developer expectations around what kinds of release versions are capable of breaking their code, and can allow [less awkward patching](https://github.com/conda-forge/conda-forge-repodata-patches-feedstock/pull/166#issuecomment-909360970) to cleanly separate compatible from incompatible versions.

We should also solicit feedback from users on how we can best warn of impending API changes in a manner that will ensure the ecosystem continues to operate smoothly.

**Comment by sroet** [comment]
Just leaving my comments on some of these things, mainly from my experience from OpenPathSampling (OPS), so cc'ing @dwhswenson:
> Since we have already released OpenMM 7.6, if a small number of users are already using this package successfully, it could potentially be problematic for reproducibility to remove it entirely. 

It did solve [this bug](https://github.com/openmm/openmm/issues/3098) for OPS. which would be reintroduced by removing the current package

> We need a more robust beta and release candidate feedback mechanism. It would be optimal if we could work with key stakeholders in the OpenMM ecosystem of downstream packages to integrate automated nightly tests of nightly builds and beta/release candidates so we can have an "early warning system" of any changes that will impact major stakeholders, allowing us ample time to work with them prior to a release that will cause major downstream issues.

OPS does have [this github actions workflow](https://github.com/openpathsampling/openpathsampling/blob/master/.github/workflows/check-openmm-rc.yml) which uses [this action](https://github.com/marketplace/actions/check-conda-for-release-candidates) to check if an `openmm` package has been uploaded to conda-forge with the `openmm_rc`  label in the last couple days. If that is true we run our test-suite against the candidate. 

This got us to [catch these issues early and fix them](https://github.com/openpathsampling/openpathsampling/pull/1051) and should be easily transferable to other stakeholders that use github actions.

**Comment by mattwthompson** [comment]
https://github.com/conda-forge/parmed-feedstock/pull/19 should help calm the waters once it is approved and merged by a feedstock maintainer. Though I will admit I said that about 3.4.2 and was wrong for not looking more closely into the details.

**Comment by dwhswenson** [comment]
> OPS does have [this github actions workflow](https://github.com/openpathsampling/openpathsampling/blob/master/.github/workflows/check-openmm-rc.yml) which uses [this action](https://github.com/marketplace/actions/check-conda-for-release-candidates) to check if an `openmm` package has been uploaded to conda-forge with the `openmm_rc` label in the last couple days. If that is true we run [our test-suite](https://github.com/openpathsampling/openpathsampling/blob/master/.github/workflows/test-openmm-rc.yml) against the candidate.


Just for reference, I put a little more detail on that GitHub action [in a blog post](http://hyperblazer.net/2021/conda-check-rc/). So far, I'm finding it to be a really useful tool -- relatively easy to set up, testing against the RC requires no thought/effort on my part, and I got notified of errors.

From my experience so far, my recommendations for any developers who want to use that action:

1. Add an if-statement to your workflow using the action so it only runs in the main repo of the project, to avoid annoying forks ([example](https://github.com/openpathsampling/openpathsampling/commit/2f6af6a79c30e2067fb9c6701468be55fbe75fd6)).
2. Run it nightly while checking a 7-day window for RCs. I found that the run every 2nd day, 3-day window I was originally using felt like it wasn't enough.

Currently, that action only checks for RCs (not betas), but it should be pretty straightforward to add support for checking *any* pre-release. If I don't get around to doing that myself, anyone else should feel free to [make a PR](https://github.com/dwhswenson/conda-rc-check) to add that.

**Comment by peastman** [comment]
> OPS does have this github actions workflow which uses this action to check if an openmm package has been uploaded to conda-forge with the openmm_rc label in the last couple days. If that is true we run our test-suite against the candidate.

That sounds really useful.  Having CI run against betas will greatly increase the chance of problems being found.

Another way of doing the same thing would be to set channel priority to `flexible` and include both `conda-forge` and `conda-forge/label/openmm_rc` as channels.  In that case, it would always test against the most recent version available, whether it's a stable release or a prerelease.  That's much simpler to implement, but changing the channel priority might cause other problems if you're also installing from additional channels.

Another question we should consider is how to make sure everyone who will be affected is involved in relevant discussions.  The AMOEBA changes were discussed in #2700.  I made sure to tag everyone involved in Tinker and Tinker-HP, since I knew it would affect them.  But I didn't realize ParmEd also included partial AMOEBA support, so I didn't know to tag anyone about it.  It's impossible for us to know every single person who is coding against OpenMM, but is there anything we can do to better reach relevant people when changes are first being discussed?

**Comment by jchodera** [comment]
Instead of serving up latest/ by literally copying all files into a latest/
folder, should we see if there's some way to redirect/rewrite URLs the
latest version number 7.6.0/ so that any copied URLs will remain static and
invariant (but pinned to that version)?

**Comment by dwhswenson** [comment]
> Can't really do much about it now, I don't think, and arguably shouldn't. Everything works fine if I start from docs.openmm.org and search around.

I don't know if anything can be done to get encourage Google to crawl and update, but if so, that would be worthwhile. I just hit search results pointing to a 404 -- I was trying to get to `app.pdbfile.PDBFile` docs, top link was to 7.5 docs (as "latest") and 404'd. Lower results pointing to 7.1 and 7.0 were valid, though.

**Comment by jchodera** [comment]
@dwhswenson : I've [requested reindexing](https://support.google.com/webmasters/answer/9012289#request_indexing) of the OpenMM website and the four docs pages. I'm not quite sure how to expire the old indexed links, though---let me know if you come up with anything!

**Comment by jchodera** [comment]
Here's another issue caused by our namespace changes: https://github.com/leeping/forcebalance/pull/239

**Comment by jchodera** [comment]
Not sure if this was mentioned earlier in the thread, but [tinker-openmm](https://github.com/TinkerTools/tinker-openmm) has a huge notice on the front page that OpenMM broke everything for them:
![image](https://user-images.githubusercontent.com/3656088/135511523-5d68ec0f-5b97-4d9f-8d69-b6f4a0ec451f.png)

**Comment by jchodera** [comment]
Another [repo bitten by the namespace change](https://github.com/aqlaboratory/openfold/issues/5).

@peastman : Can you comment on the [medium and long term changes to our release process proposed here](https://github.com/openmm/openmm/issues/3232#issue-984050206)? I want to make sure we are able to minimize disruptions for the upcoming OpenMM 7.7 release.

**Comment by jchodera** [comment]
Three more codes impacted by the namespace change---I'm reaching out to try to help them upgrade:
* [mdtraj/mdtraj](https://github.com/mdtraj/mdtraj/issues/1695)
* [choderalab/fah-xchem](https://github.com/choderalab/fah-xchem/issues/160)
* [deepmind/alphafold](https://github.com/deepmind/alphafold/blob/cef198e0b650bf35431419db1d91d6aedac75e82/docker/Dockerfile#L55)

**Comment by peastman** [comment]
I don't think it's particularly causing problems for them.  MDTraj is working fine, just getting a deprecation warning (as intended).  AlphaFold was already pinning to 7.5.1 so it won't be affected until they choose to upgrade (at which point it will probably also continue working, just getting a deprecation warning).

**Comment by dotsdl** [comment]
See [my comment](https://github.com/mdtraj/mdtraj/issues/1695#issuecomment-972374270) re: mdtraj. I believe I barked up the wrong tree too hastily. Apologies for the noise!

@jchodera I am seeing many uses of `simtk` in `perses` `master`; in particular seeing the import we see complaints for from `fah-xchem` in `dev`:

```
No module named 'simtk.openmm.app.topology'
```

Looks like we already have [an issue](https://github.com/choderalab/perses/issues/837) for this.
